[PyTorch ] Thread parallel bmm across batch dim

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#59596 [PyTorch ] Thread parallel bmm across batch dim**
* #59595 [Pytorch benchmark] Add BMM benchmark

Parallelize batch matmul across batch dim. This was found to improve perf for
some usecases on mobile.

Benchmarking results:
BMM benchmark via operator benchmark on Samsung S8 US phone:



B | M | N | K | &nbsp; | No Threading | Threads=4
-- | -- | -- | -- | -- | -- | --
2 | 8 | 256 | 16 | 8448 | 0.062 | 0.095
2 | 8 | 256 | 32 | 16896 | 0.077 | 0.099
2 | 8 | 16 | 16 | 768 | 0.048 | 0.07
2 | 8 | 16 | 32 | 1536 | 0.048 | 0.075
2 | 256 | 256 | 16 | 16384 | 0.476 | 0.27
2 | 256 | 256 | 32 | 32768 | 0.79 | 0.477
2 | 256 | 16 | 16 | 8704 | 0.08 | 0.101
2 | 256 | 16 | 32 | 17408 | 0.11 | 0.134
100 | 8 | 256 | 16 | 422400 | 2.269 | 0.6945
100 | 8 | 256 | 32 | 844800 | 3.111 | 0.92
100 | 8 | 16 | 16 | 38400 | 0.936 | 0.425
100 | 8 | 16 | 32 | 76800 | 0.987 | 0.45
100 | 256 | 256 | 16 | 819200 | 27.601 | 10.7545
100 | 256 | 256 | 32 | 1638400 | 43.088 | 13.756
100 | 256 | 16 | 16 | 435200 | 2.734 | 0.802
100 | 256 | 16 | 32 | 870400 | 3.9125 | 1.174
&nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp;
4 | 32 | 32 | 16 | 4096 | 0.082 | 0.08
4 | 32 | 32 | 32 | 8192 | 0.091 | 0.082
4 | 32 | 64 | 16 | 6144 | 0.092 | 0.101
4 | 32 | 64 | 32 | 12288 | 0.114 | 0.093
4 | 64 | 32 | 16 | 6144 | 0.0985 | 0.0845
4 | 64 | 32 | 32 | 12288 | 0.116 | 0.101
4 | 64 | 64 | 16 | 8192 | 0.119 | 0.093
4 | 64 | 64 | 32 | 16384 | 0.161 | 0.127
8 | 32 | 32 | 16 | 8192 | 0.129 | 0.1
8 | 32 | 32 | 32 | 16384 | 0.153 | 0.103
8 | 32 | 64 | 16 | 12288 | 0.155 | 0.117
8 | 32 | 64 | 32 | 24576 | 0.199 | 0.1465
8 | 64 | 32 | 16 | 12288 | 0.159 | 0.116
8 | 64 | 32 | 32 | 24576 | 0.204 | 0.138
8 | 64 | 64 | 16 | 16384 | 0.209 | 0.156
8 | 64 | 64 | 32 | 32768 | 0.292 | 0.256


Seems that even at smaller size we are almost on par or little better. Only small batch size (2) seems to be worse.



B | M | N | K | &nbsp; | No Threading | Threads=4 | Speedup
-- | -- | -- | -- | -- | -- | -- | --
4 | 4 | 16 | 16 | 1280 | 0.068 | 0.065 | 1.04615385
4 | 4 | 32 | 32 | 4608 | 0.068 | 0.07 | 0.97142857
4 | 4 | 32 | 64 | 9216 | 0.072 | 0.072 | 1
4 | 4 | 64 | 32 | 8704 | 0.073 | 0.076 | 0.96052632
4 | 4 | 64 | 64 | 17408 | 0.083 | 0.08 | 1.0375
4 | 8 | 32 | 32 | 5120 | 0.083 | 0.072 | 1.15277778
4 | 8 | 32 | 64 | 10240 | 0.079 | 0.079 | 1
4 | 8 | 64 | 32 | 9216 | 0.091 | 0.075 | 1.21333333
4 | 8 | 64 | 64 | 18432 | 0.14 | 0.075 | 1.86666667











Differential Revision: [D26833417](https://our.internmc.facebook.com/intern/diff/D26833417/)

kimishpatel committed:
[PyTorch ] Thread parallel bmm across batch dim

Parallelize batch matmul across batch dim. This was found to improve perf for
some usecases on mobile.

Differential Revision: [D26833417](https://our.internmc.facebook.com/intern/diff/D26833417/)

[ghstack-poisoned]

ailzhang commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1257,13 +1257,49 @@ static inline Tensor& bmm_out_or_baddbmm_(Tensor& self_or_result, const Tensor&
-        auto r = self_or_result.select(0, b);
-        addmm_impl_cpu_(r, r, batch1.select(0, b), batch2.select(0, b), 0, 1);
+      /*
+       * Inference mode multithreading is done because various thread local
```
Would you mind also opening an issue and documenting what's the new API would do there? :D

ailzhang replied:
Also I'd appreciate a big warning here: if someone wants to add another instance of this they should do it properly by using the new API. :D

ailzhang reviewed (approved):
Stamping since this PR guarantees correctness, although we should keep in mind `at::parallel_for` in its current shape doesn't support preserving TLS state. So this hack should be discouraged in new call sites.

albanD commented:
The BLAS library that is called inside `addmm_impl_cpu_` is already thread parallelism on CPU. Is it possible that this would lead to oversubscription of threads on machines with multithreaded BLAS (and so big slow down)?

Also is this issue due to the fact that the BLAS used on mobile is not multithreaded? If that's the case, wouldn't we prefer to add this to this BLAS only and not at such a high level?

kimishpatel commented:
> The BLAS library that is called inside addmm_impl_cpu_ is already thread parallelism on CPU. Is it possible that this would lead to oversubscription of threads on machines with multithreaded BLAS (and so big slow down)?

@albanD is this for oss build? If so can you point me to cmake config? We can disable this if you want for server side. Also we most likely use MKL path. On the other hand however having BLAS have its own threadpool along with pytorch its own threadpool will already be detrimental to performance.

> Also is this issue due to the fact that the BLAS used on mobile is not multithreaded? If that's the case, wouldn't we prefer to add this to this BLAS only and not at such a high level?

On mobile side yes we dont use BLAS's multithreading for the same reason that too many threads can cause perf regression as threads may compete for scheduling.

albanD commented:
> is this for oss build?

At least yes, I don't know which BLAS lib is used internally.

> If so can you point me to cmake config?

I am not sure what you're looking for. The user control that parallelism with `OMP_NUM_THREADS` env variable or https://pytorch.org/docs/stable/generated/torch.set_num_threads.html?highlight=set_num_threads#torch.set_num_threads.

> Also we most likely use MKL path.

That depends on what you compiled with.
But even when MKL is used, the codeblock above will only be used in a very narrow case and the code you are changing will be called otherwise (calling again into the non-batched version of MKL).

> On the other hand however having BLAS have its own threadpool along with pytorch its own threadpool will already be detrimental to performance.

That's a know issue. But in practice, the pytorch threadpool is rarely used and so the BLAS threadpool is the only one doing actual work.

> On mobile side yes we dont use BLAS's multithreading for the same reason that too many threads can cause perf regression as threads may compete for scheduling.

Would you be able to make this change inside the mobile blas binding directly to avoid having divergent callpath for mobile/non-mobile in this high level function?

albanD commented:
Ho sorry,
While checking the code, the `addmm_impl_cpu_` function that you call end up calling`at::native::cpublas::gemm`, that calls into `dgemm_` from whatever was the blas library was available at compile time (or a manual implementation if nothing is provided).

I was wondering if there isn't a better place to put this code in this stack. In particular, if you do it lower (when there are no more Tensor ops), you don't need all the special logic to handle the fact that parallel_for does not propagate TLS.

kimishpatel commented:
> Ho sorry,
> While checking the code, the `addmm_impl_cpu_` function that you call end up calling`at::native::cpublas::gemm`, that calls into `dgemm_` from whatever was the blas library was available at compile time (or a manual implementation if nothing is provided).
> 
> I was wondering if there isn't a better place to put this code in this stack. In particular, if you do it lower (when there are no more Tensor ops), you don't need all the special logic to handle the fact that parallel_for does not propagate TLS.

I tried implementation which does not require propagation of TLS in D28873227. But it excludes other things that can be parallelized. For example select op etc. This results in cutting the benefit in half.

Also from the discussion it seems that we dont want paralllelization on non-mobile builds.

albanD commented:
> Also from the discussion it seems that we dont want paralllelization on non-mobile builds.

I don't think we do, but that would still simplify the mobile only code :)

kimishpatel commented:
> > Also from the discussion it seems that we dont want paralllelization on non-mobile builds.
> 
> I don't think we do, but that would still simplify the mobile only code :)

So I can make this mobile only? But what would still simplify the mobile only code? As I said doing this at lower level is not gonna be perf win as much.

kimishpatel committed:
Update on "[PyTorch ] Thread parallel bmm across batch dim"


Parallelize batch matmul across batch dim. This was found to improve perf for
some usecases on mobile.

Differential Revision: [D26833417](https://our.internmc.facebook.com/intern/diff/D26833417/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[PyTorch ] Thread parallel bmm across batch dim"


Parallelize batch matmul across batch dim. This was found to improve perf for
some usecases on mobile.

Benchmarking results:
BMM benchmark via operator benchmark on Samsung S8 US phone:



B | M | N | K | &nbsp; | No Threading | Threads=4
-- | -- | -- | -- | -- | -- | --
2 | 8 | 256 | 16 | 8448 | 0.062 | 0.095
2 | 8 | 256 | 32 | 16896 | 0.077 | 0.099
2 | 8 | 16 | 16 | 768 | 0.048 | 0.07
2 | 8 | 16 | 32 | 1536 | 0.048 | 0.075
2 | 256 | 256 | 16 | 16384 | 0.476 | 0.27
2 | 256 | 256 | 32 | 32768 | 0.79 | 0.477
2 | 256 | 16 | 16 | 8704 | 0.08 | 0.101
2 | 256 | 16 | 32 | 17408 | 0.11 | 0.134
100 | 8 | 256 | 16 | 422400 | 2.269 | 0.6945
100 | 8 | 256 | 32 | 844800 | 3.111 | 0.92
100 | 8 | 16 | 16 | 38400 | 0.936 | 0.425
100 | 8 | 16 | 32 | 76800 | 0.987 | 0.45
100 | 256 | 256 | 16 | 819200 | 27.601 | 10.7545
100 | 256 | 256 | 32 | 1638400 | 43.088 | 13.756
100 | 256 | 16 | 16 | 435200 | 2.734 | 0.802
100 | 256 | 16 | 32 | 870400 | 3.9125 | 1.174
&nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp;
4 | 32 | 32 | 16 | 4096 | 0.082 | 0.08
4 | 32 | 32 | 32 | 8192 | 0.091 | 0.082
4 | 32 | 64 | 16 | 6144 | 0.092 | 0.101
4 | 32 | 64 | 32 | 12288 | 0.114 | 0.093
4 | 64 | 32 | 16 | 6144 | 0.0985 | 0.0845
4 | 64 | 32 | 32 | 12288 | 0.116 | 0.101
4 | 64 | 64 | 16 | 8192 | 0.119 | 0.093
4 | 64 | 64 | 32 | 16384 | 0.161 | 0.127
8 | 32 | 32 | 16 | 8192 | 0.129 | 0.1
8 | 32 | 32 | 32 | 16384 | 0.153 | 0.103
8 | 32 | 64 | 16 | 12288 | 0.155 | 0.117
8 | 32 | 64 | 32 | 24576 | 0.199 | 0.1465
8 | 64 | 32 | 16 | 12288 | 0.159 | 0.116
8 | 64 | 32 | 32 | 24576 | 0.204 | 0.138
8 | 64 | 64 | 16 | 16384 | 0.209 | 0.156
8 | 64 | 64 | 32 | 32768 | 0.292 | 0.256








Differential Revision: [D26833417](https://our.internmc.facebook.com/intern/diff/D26833417/)

[ghstack-poisoned]

albanD commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1256,14 +1256,64 @@ static inline Tensor& bmm_out_or_baddbmm_(Tensor& self_or_result, const Tensor&
+    // bmm_test: operator benchmark under
+    // benchmarks/operator_benchmarks/pt/bmm_test.py Ran this benchmark for
+    // various matrix sizes on Samsung S8U
+    enable_multithreaded_bmm =
```
You want to use `&=` here no?

kimishpatel replied:
Oh yes. Thanks for the catch.

albanD commented on test/test_linalg.py:
```diff
@@ -6196,6 +6196,58 @@ def generate_inputs(num_batches):
 
+    @onlyCPU
+    @torch.inference_mode()
+    def test_bmm_multithreaded(self, device):
```
This file is actually only about `torch.linalg.*` functions. Could you move this test to `test_torch.py`?

kimishpatel replied:
Why does it make sense for it to be in test_torch?

albanD replied:
Everything that has no better place goes there. In particular testing for all operators that doesn't fit in any better category goes there.

albanD commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1256,14 +1256,64 @@ static inline Tensor& bmm_out_or_baddbmm_(Tensor& self_or_result, const Tensor&
             && self_or_result.is_contiguous()) {
     at::native::_baddbmm_mkl_(self_or_result, batch1, batch2, beta, alpha);
   } else { // split along batch dimension
+    bool enable_multithreaded_bmm{false};
```
Could you move this into the #else of the macro and make it const?
That would allow the compiler to completely prune the code in the conditions below: https://godbolt.org/z/8894Mdxd1

kimishpatel replied:
Oh interesting. I did not realize that without `const` compiler would not 'const propagate'. Seems strange. But sure I can make the change.

kimishpatel replied:
If you do "-Os" then it simplifies a lot. Nonetheless I will make the change.

albanD replied:
It does with optimization enabled. But not for debug versions from what I see.

albanD reviewed (commented):
Thanks for the update.
I think we can remove a bit of code duplication but the overall idea (except the &=) looks ok to me.

kimishpatel commented on test/test_linalg.py:
```diff
@@ -6196,6 +6196,58 @@ def generate_inputs(num_batches):
 
+    @onlyCPU
+    @torch.inference_mode()
+    def test_bmm_multithreaded(self, device):
```
I didnt quite check that. Can report back once I do.

albanD replied:
If it takes more than 15s, we should consider marking it as a slow test.

kimishpatel commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1256,14 +1256,64 @@ static inline Tensor& bmm_out_or_baddbmm_(Tensor& self_or_result, const Tensor&
-        auto r = self_or_result.select(0, b);
-        addmm_impl_cpu_(r, r, batch1.select(0, b), batch2.select(0, b), 0, 1);
+      if (enable_multithreaded_bmm) {
+        auto bmm_out_fn = [&](uint64_t start, uint64_t end) {
```
Good suggestion.

kimishpatel replied:
Actually just realized that we do InferenceMode guard in the lambda so cannot quite replace it as per your suggestion.

albanD replied:
That's unfortunate.. I guess that adding another condition there is going to be annoying.. ok

kimishpatel commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1256,14 +1256,64 @@ static inline Tensor& bmm_out_or_baddbmm_(Tensor& self_or_result, const Tensor&
+    bool enable_multithreaded_bmm{false};
+#ifdef C10_MOBILE
+    /*
+     * Inference mode multithreading is done because various thread local
```
Not sure what you mean. It already clarifies why it is enabled in inference mode. Or do you mean to capture the fact that having autograd engine running in multi-threading can be an issue?

albanD replied:
It is more that "Inference mode multithreading is done because" is not clear to me at first read. I would replace it with "We only do multithreading when Inference Mode is enabled because".

kimishpatel committed:
Update on "[PyTorch ] Thread parallel bmm across batch dim"


Parallelize batch matmul across batch dim. This was found to improve perf for
some usecases on mobile.

Benchmarking results:
BMM benchmark via operator benchmark on Samsung S8 US phone:



B | M | N | K | &nbsp; | No Threading | Threads=4
-- | -- | -- | -- | -- | -- | --
2 | 8 | 256 | 16 | 8448 | 0.062 | 0.095
2 | 8 | 256 | 32 | 16896 | 0.077 | 0.099
2 | 8 | 16 | 16 | 768 | 0.048 | 0.07
2 | 8 | 16 | 32 | 1536 | 0.048 | 0.075
2 | 256 | 256 | 16 | 16384 | 0.476 | 0.27
2 | 256 | 256 | 32 | 32768 | 0.79 | 0.477
2 | 256 | 16 | 16 | 8704 | 0.08 | 0.101
2 | 256 | 16 | 32 | 17408 | 0.11 | 0.134
100 | 8 | 256 | 16 | 422400 | 2.269 | 0.6945
100 | 8 | 256 | 32 | 844800 | 3.111 | 0.92
100 | 8 | 16 | 16 | 38400 | 0.936 | 0.425
100 | 8 | 16 | 32 | 76800 | 0.987 | 0.45
100 | 256 | 256 | 16 | 819200 | 27.601 | 10.7545
100 | 256 | 256 | 32 | 1638400 | 43.088 | 13.756
100 | 256 | 16 | 16 | 435200 | 2.734 | 0.802
100 | 256 | 16 | 32 | 870400 | 3.9125 | 1.174
&nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp;
4 | 32 | 32 | 16 | 4096 | 0.082 | 0.08
4 | 32 | 32 | 32 | 8192 | 0.091 | 0.082
4 | 32 | 64 | 16 | 6144 | 0.092 | 0.101
4 | 32 | 64 | 32 | 12288 | 0.114 | 0.093
4 | 64 | 32 | 16 | 6144 | 0.0985 | 0.0845
4 | 64 | 32 | 32 | 12288 | 0.116 | 0.101
4 | 64 | 64 | 16 | 8192 | 0.119 | 0.093
4 | 64 | 64 | 32 | 16384 | 0.161 | 0.127
8 | 32 | 32 | 16 | 8192 | 0.129 | 0.1
8 | 32 | 32 | 32 | 16384 | 0.153 | 0.103
8 | 32 | 64 | 16 | 12288 | 0.155 | 0.117
8 | 32 | 64 | 32 | 24576 | 0.199 | 0.1465
8 | 64 | 32 | 16 | 12288 | 0.159 | 0.116
8 | 64 | 32 | 32 | 24576 | 0.204 | 0.138
8 | 64 | 64 | 16 | 16384 | 0.209 | 0.156
8 | 64 | 64 | 32 | 32768 | 0.292 | 0.256








Differential Revision: [D26833417](https://our.internmc.facebook.com/intern/diff/D26833417/)

[ghstack-poisoned]

albanD reviewed (approved):
LGTM

We might want to add the slowtest on top of that if it takes too long to run though.

kimishpatel committed:
Update on "[PyTorch ] Thread parallel bmm across batch dim"


Parallelize batch matmul across batch dim. This was found to improve perf for
some usecases on mobile.

Benchmarking results:
BMM benchmark via operator benchmark on Samsung S8 US phone:



B | M | N | K | &nbsp; | No Threading | Threads=4
-- | -- | -- | -- | -- | -- | --
2 | 8 | 256 | 16 | 8448 | 0.062 | 0.095
2 | 8 | 256 | 32 | 16896 | 0.077 | 0.099
2 | 8 | 16 | 16 | 768 | 0.048 | 0.07
2 | 8 | 16 | 32 | 1536 | 0.048 | 0.075
2 | 256 | 256 | 16 | 16384 | 0.476 | 0.27
2 | 256 | 256 | 32 | 32768 | 0.79 | 0.477
2 | 256 | 16 | 16 | 8704 | 0.08 | 0.101
2 | 256 | 16 | 32 | 17408 | 0.11 | 0.134
100 | 8 | 256 | 16 | 422400 | 2.269 | 0.6945
100 | 8 | 256 | 32 | 844800 | 3.111 | 0.92
100 | 8 | 16 | 16 | 38400 | 0.936 | 0.425
100 | 8 | 16 | 32 | 76800 | 0.987 | 0.45
100 | 256 | 256 | 16 | 819200 | 27.601 | 10.7545
100 | 256 | 256 | 32 | 1638400 | 43.088 | 13.756
100 | 256 | 16 | 16 | 435200 | 2.734 | 0.802
100 | 256 | 16 | 32 | 870400 | 3.9125 | 1.174
&nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp;
4 | 32 | 32 | 16 | 4096 | 0.082 | 0.08
4 | 32 | 32 | 32 | 8192 | 0.091 | 0.082
4 | 32 | 64 | 16 | 6144 | 0.092 | 0.101
4 | 32 | 64 | 32 | 12288 | 0.114 | 0.093
4 | 64 | 32 | 16 | 6144 | 0.0985 | 0.0845
4 | 64 | 32 | 32 | 12288 | 0.116 | 0.101
4 | 64 | 64 | 16 | 8192 | 0.119 | 0.093
4 | 64 | 64 | 32 | 16384 | 0.161 | 0.127
8 | 32 | 32 | 16 | 8192 | 0.129 | 0.1
8 | 32 | 32 | 32 | 16384 | 0.153 | 0.103
8 | 32 | 64 | 16 | 12288 | 0.155 | 0.117
8 | 32 | 64 | 32 | 24576 | 0.199 | 0.1465
8 | 64 | 32 | 16 | 12288 | 0.159 | 0.116
8 | 64 | 32 | 32 | 24576 | 0.204 | 0.138
8 | 64 | 64 | 16 | 16384 | 0.209 | 0.156
8 | 64 | 64 | 32 | 32768 | 0.292 | 0.256








Differential Revision: [D26833417](https://our.internmc.facebook.com/intern/diff/D26833417/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[PyTorch ] Thread parallel bmm across batch dim"


Parallelize batch matmul across batch dim. This was found to improve perf for
some usecases on mobile.

Benchmarking results:
BMM benchmark via operator benchmark on Samsung S8 US phone:



B | M | N | K | &nbsp; | No Threading | Threads=4
-- | -- | -- | -- | -- | -- | --
2 | 8 | 256 | 16 | 8448 | 0.062 | 0.095
2 | 8 | 256 | 32 | 16896 | 0.077 | 0.099
2 | 8 | 16 | 16 | 768 | 0.048 | 0.07
2 | 8 | 16 | 32 | 1536 | 0.048 | 0.075
2 | 256 | 256 | 16 | 16384 | 0.476 | 0.27
2 | 256 | 256 | 32 | 32768 | 0.79 | 0.477
2 | 256 | 16 | 16 | 8704 | 0.08 | 0.101
2 | 256 | 16 | 32 | 17408 | 0.11 | 0.134
100 | 8 | 256 | 16 | 422400 | 2.269 | 0.6945
100 | 8 | 256 | 32 | 844800 | 3.111 | 0.92
100 | 8 | 16 | 16 | 38400 | 0.936 | 0.425
100 | 8 | 16 | 32 | 76800 | 0.987 | 0.45
100 | 256 | 256 | 16 | 819200 | 27.601 | 10.7545
100 | 256 | 256 | 32 | 1638400 | 43.088 | 13.756
100 | 256 | 16 | 16 | 435200 | 2.734 | 0.802
100 | 256 | 16 | 32 | 870400 | 3.9125 | 1.174
&nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp; | &nbsp;
4 | 32 | 32 | 16 | 4096 | 0.082 | 0.08
4 | 32 | 32 | 32 | 8192 | 0.091 | 0.082
4 | 32 | 64 | 16 | 6144 | 0.092 | 0.101
4 | 32 | 64 | 32 | 12288 | 0.114 | 0.093
4 | 64 | 32 | 16 | 6144 | 0.0985 | 0.0845
4 | 64 | 32 | 32 | 12288 | 0.116 | 0.101
4 | 64 | 64 | 16 | 8192 | 0.119 | 0.093
4 | 64 | 64 | 32 | 16384 | 0.161 | 0.127
8 | 32 | 32 | 16 | 8192 | 0.129 | 0.1
8 | 32 | 32 | 32 | 16384 | 0.153 | 0.103
8 | 32 | 64 | 16 | 12288 | 0.155 | 0.117
8 | 32 | 64 | 32 | 24576 | 0.199 | 0.1465
8 | 64 | 32 | 16 | 12288 | 0.159 | 0.116
8 | 64 | 32 | 32 | 24576 | 0.204 | 0.138
8 | 64 | 64 | 16 | 16384 | 0.209 | 0.156
8 | 64 | 64 | 32 | 32768 | 0.292 | 0.256








Differential Revision: [D26833417](https://our.internmc.facebook.com/intern/diff/D26833417/)

[ghstack-poisoned]
